### PR TITLE
Introduce the linkerd.io/shared-component label

### PR DESCRIPTION
**Pull request by ihcsim**
Following https://github.com/linkerd/linkerd2/issues/2376#issuecomment-506793743, this PR introduces a new label named, `linkerd.io/shared-component`, which can be used to retrieve all Linkerd shared and non-shared global resources, as well as namespace-scoped resources. The reason why this is a better choice than `linkerd.io/control-plane-ns` is because it doesn't imply the namespace served by shared global resources like CRDs.

`linkerd check --pre` is also updated to use this label to check for existence of global resources.

```
k -n linkerd get all,configmap,secret,sa,psp,crd,clusterrole,clusterrolebinding,role,rolebinding,mutatingwebhookconfiguration,validatingwebhookconfiguration -l linkerd.io/shared-component -L linkerd.io/shared-component
NAME                             TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)             AGE     SHARED-COMPONENT
service/linkerd-controller-api   ClusterIP   10.99.251.216    <none>        8085/TCP            6m48s   false
service/linkerd-destination      ClusterIP   10.104.177.185   <none>        8086/TCP            6m48s   false
service/linkerd-grafana          ClusterIP   10.98.137.78     <none>        3000/TCP            6m48s   false
service/linkerd-identity         ClusterIP   10.108.195.71    <none>        8080/TCP            6m48s   false
service/linkerd-prometheus       ClusterIP   10.110.243.151   <none>        9090/TCP            6m48s   false
service/linkerd-proxy-injector   ClusterIP   10.101.199.147   <none>        443/TCP             6m48s   false
service/linkerd-sp-validator     ClusterIP   10.99.240.44     <none>        443/TCP             6m47s   false
service/linkerd-tap              ClusterIP   10.98.184.220    <none>        8088/TCP            6m47s   false
service/linkerd-web              ClusterIP   10.100.112.100   <none>        8084/TCP,9994/TCP   6m48s   false

NAME                                     READY   UP-TO-DATE   AVAILABLE   AGE     SHARED-COMPONENT
deployment.apps/linkerd-controller       1/1     1            1           6m48s   false
deployment.apps/linkerd-grafana          1/1     1            1           6m48s   false
deployment.apps/linkerd-identity         1/1     1            1           6m48s   false
deployment.apps/linkerd-prometheus       1/1     1            1           6m48s   false
deployment.apps/linkerd-proxy-injector   1/1     1            1           6m48s   false
deployment.apps/linkerd-sp-validator     1/1     1            1           6m47s   false
deployment.apps/linkerd-tap              1/1     1            1           6m47s   false
deployment.apps/linkerd-web              1/1     1            1           6m48s   false

NAME                                  DATA   AGE     SHARED-COMPONENT
configmap/linkerd-config              3      6m48s   false
configmap/linkerd-grafana-config      3      6m48s   false
configmap/linkerd-prometheus-config   1      6m48s   false

NAME                                TYPE     DATA   AGE     SHARED-COMPONENT
secret/linkerd-identity-issuer      Opaque   2      6m48s   false
secret/linkerd-proxy-injector-tls   Opaque   2      6m49s   false
secret/linkerd-sp-validator-tls     Opaque   2      6m48s   false

NAME                                    SECRETS   AGE     SHARED-COMPONENT
serviceaccount/linkerd-controller       1         6m49s   false
serviceaccount/linkerd-grafana          1         6m49s   false
serviceaccount/linkerd-identity         1         6m49s   false
serviceaccount/linkerd-prometheus       1         6m49s   false
serviceaccount/linkerd-proxy-injector   1         6m49s   false
serviceaccount/linkerd-sp-validator     1         6m49s   false
serviceaccount/linkerd-tap              1         6m48s   false
serviceaccount/linkerd-web              1         6m49s   false

NAME                                                         PRIV    CAPS                SELINUX    RUNASUSER   FSGROUP     SUPGROUP    READONLYROOTFS   VOLUMES                                                                 SHARED-COMPONENT
podsecuritypolicy.extensions/linkerd-linkerd-control-plane   false   NET_ADMIN,NET_RAW   RunAsAny   RunAsAny    MustRunAs   MustRunAs   true             configMap,emptyDir,secret,projected,downwardAPI,persistentVolumeClaim   false

NAME                                                                       CREATED AT             SHARED-COMPONENT
customresourcedefinition.apiextensions.k8s.io/serviceprofiles.linkerd.io   2019-06-28T18:48:35Z   true

NAME                                                                   AGE     SHARED-COMPONENT
clusterrole.rbac.authorization.k8s.io/linkerd-linkerd-controller       6m48s   false
clusterrole.rbac.authorization.k8s.io/linkerd-linkerd-identity         6m48s   false
clusterrole.rbac.authorization.k8s.io/linkerd-linkerd-prometheus       6m48s   false
clusterrole.rbac.authorization.k8s.io/linkerd-linkerd-proxy-injector   6m48s   false
clusterrole.rbac.authorization.k8s.io/linkerd-linkerd-sp-validator     6m48s   false
clusterrole.rbac.authorization.k8s.io/linkerd-linkerd-tap              6m47s   false

NAME                                                                          AGE     SHARED-COMPONENT
clusterrolebinding.rbac.authorization.k8s.io/linkerd-linkerd-controller       6m48s   false
clusterrolebinding.rbac.authorization.k8s.io/linkerd-linkerd-identity         6m48s   false
clusterrolebinding.rbac.authorization.k8s.io/linkerd-linkerd-prometheus       6m48s   false
clusterrolebinding.rbac.authorization.k8s.io/linkerd-linkerd-proxy-injector   6m48s   false
clusterrolebinding.rbac.authorization.k8s.io/linkerd-linkerd-sp-validator     6m48s   false
clusterrolebinding.rbac.authorization.k8s.io/linkerd-linkerd-tap              6m47s   false

NAME                                         AGE     SHARED-COMPONENT
role.rbac.authorization.k8s.io/linkerd-psp   6m47s   false

NAME                                                AGE     SHARED-COMPONENT
rolebinding.rbac.authorization.k8s.io/linkerd-psp   6m47s   false

NAME                                                                                              CREATED AT             SHARED-COMPONENT
mutatingwebhookconfiguration.admissionregistration.k8s.io/linkerd-proxy-injector-webhook-config   2019-06-28T18:48:35Z   false

NAME                                                                                              CREATED AT             SHARED-COMPONENT
validatingwebhookconfiguration.admissionregistration.k8s.io/linkerd-sp-validator-webhook-config   2019-06-28T18:48:36Z   false
```

### CODE REVIEW

File: bin/test-cleanup [hunk 1]
```diff
@@ -12,31 +12,31 @@ fi
 
 echo "cleaning up namespace [${linkerd_namespace}] in k8s-context [${k8s_context}] and associated test namespaces"
 
-if ! namespaces=$(kubectl --context=$k8s_context get ns -oname | grep -E "/$linkerd_namespace(-|$)"); then
+if ! namespaces=$(kubectl --context=$k8s_context get ns -l linkerd.io/shared-component -oname); then
```
Comment (ihcsim) [inline bin/test-cleanup]: I think this integration test clean-up script can be simplified further, in another PR.

**Comment by ihcsim** [comment]
@siggy Thanks for the feedback. I can see how others can get confused by this new label. Basically, we use the `linkerd.io/shared-component` to distinguish global resources that will remain as singletons (hence, _shared), in a multi control plane setup, from those that are bound to a control plane namespace. E.g. cluster roles are bound to the control plane namespace via bindings, mwc via its `Service` endpoint configuration etc. So far, the CRDs are the ones that will remain as singletons, hence, they are denoted with `linkerd.io/shared-component: true`.

As far as the `check --pre` and `install` commands are concerned, we just needed a label that can be used to retrieve all global (shared and unshared) and namespace-scoped resources. LMK if we deem there are other more appropriate labels, or if prefixing it with `_` makes it clearer etc.

**Comment by adleong** [comment]
As a follow-up, it would be great if we had a doc page on the website which was a glossary of linkerd's labels/annotations with their meaning and how they are used.

**Comment by grampelberg** [comment]
@adleong mind creating an issue on linkerd/website?

**Comment by adleong** [comment]
https://github.com/linkerd/website/issues/391

**Comment by ihcsim** [comment]
Closing until further decision on multi control plane setup.

**Pull request closed without merging**